core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/chain/StopStartUpscaleFlowEventChainFactory.java

```diff
@@ -0,0 +1,76 @@
+package com.sequenceiq.cloudbreak.core.flow2.chain;
+
+import static com.sequenceiq.cloudbreak.core.flow2.stack.sync.StackSyncEvent.STACK_SYNC_EVENT;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.cloudbreak.common.exception.NotFoundException;
+import com.sequenceiq.cloudbreak.common.type.ScalingType;
+import com.sequenceiq.cloudbreak.core.flow2.cluster.stopstartus.StopStartUpscaleEvent;
+import com.sequenceiq.cloudbreak.core.flow2.event.StopStartUpscaleTriggerEvent;
+import com.sequenceiq.cloudbreak.core.flow2.event.StackAndClusterUpscaleTriggerEvent;
+import com.sequenceiq.cloudbreak.core.flow2.event.StackSyncTriggerEvent;
+import com.sequenceiq.cloudbreak.domain.stack.cluster.host.HostGroup;
+import com.sequenceiq.cloudbreak.domain.view.ClusterView;
+import com.sequenceiq.cloudbreak.domain.view.StackView;
+import com.sequenceiq.cloudbreak.service.hostgroup.HostGroupService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.flow.core.chain.FlowEventChainFactory;
+import com.sequenceiq.flow.core.chain.config.FlowTriggerEventQueue;
+
+@Component
+public class StopStartUpscaleFlowEventChainFactory implements FlowEventChainFactory<StackAndClusterUpscaleTriggerEvent> {
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private HostGroupService hostGroupService;
+
+    @Override
+    public String initEvent() {
+        return FlowChainTriggers.STOPSTART_UPSCALE_CHAIN_TRIGGER_EVENT;
+    }
+
+    @Override
+    public FlowTriggerEventQueue createFlowTriggerEventQueue(StackAndClusterUpscaleTriggerEvent event) {
+        StackView stackView = stackService.getViewByIdWithoutAuth(event.getResourceId());
+        ClusterView clusterView = stackView.getClusterView();
+        Queue<Selectable> flowEventChain = new ConcurrentLinkedQueue<>();
+        addStackSyncTriggerEvent(event, flowEventChain);
+        addClusterScaleTriggerEventIfNeeded(event, stackView, clusterView, flowEventChain);
+        return new FlowTriggerEventQueue(getName(), event, flowEventChain);
+    }
+
+    // TODO CB-14929: Is a stack sync actually required?
+    private void addStackSyncTriggerEvent(StackAndClusterUpscaleTriggerEvent event, Queue<Selectable> flowEventChain) {
+        flowEventChain.add(new StackSyncTriggerEvent(
+                STACK_SYNC_EVENT.event(),
+                event.getResourceId(),
+                false,
+                event.accepted())
+        );
+    }
+
+    private void addClusterScaleTriggerEventIfNeeded(StackAndClusterUpscaleTriggerEvent event, StackView stackView, ClusterView clusterView,
+            Queue<Selectable> flowEventChain) {
+        if (ScalingType.isClusterUpScale(event.getScalingType()) && clusterView != null) {
```

Comment:
does it makes sense to startstopupscale without upscale?

Reply:
What do you suggest?

Reply:
I think we can remove ScalingType.isClusterUpScale(event.getScalingType()), but it does not harm, just commented it because I think this condition can be removed.

Reply:
Removing in the next PR.